### PR TITLE
bpo-43908: check_set_special_type_attr() and type_set_annotations() now check for immutable flag

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue43908

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @vstinner

File: Objects/typeobject.c [hunk 1]
```diff
@@ -461,7 +461,7 @@ static PyMemberDef type_members[] = {
 static int
 check_set_special_type_attr(PyTypeObject *type, PyObject *value, const char *name)
 {
-    if (!(type->tp_flags & Py_TPFLAGS_HEAPTYPE)) {
+    if (_PyType_HasFeature(type, Py_TPFLAGS_IMMUTABLETYPE)) {
         PyErr_Format(PyExc_TypeError,
                      "can't set %s.%s", type->tp_name, name);
```
Comment (vstinner) [inline Objects/typeobject.c]: I suggest to advertize the new Py_TPFLAGS_IMMUTABLETYPE flag in the error message ;-)

```suggestion
                     "can't set '%r' attribute of immutable type %s", name, type->tp_name);
```

Same remark for type_set_annotations(). Please also change type_setattro() error message.

Comment (erlend-aasland) [reply Objects/typeobject.c]: Will do!

Comment (erlend-aasland) [reply Objects/typeobject.c]: Improved comments in fd929c014bf0620748c25c32378505db69fde58f :)

File: Objects/typeobject.c [hunk 2]
```diff
@@ -3947,7 +3951,7 @@ type_setattro(PyTypeObject *type, PyObject *name, PyObject *value)
     if (type->tp_flags & Py_TPFLAGS_IMMUTABLETYPE) {
         PyErr_Format(
             PyExc_TypeError,
-            "can't set attributes of built-in/extension type '%s'",
+            "cannot set attributes of immutable type '%s'",
```
Comment (vstinner) [inline Objects/typeobject.c]: Can you please put the attribute name in the error message?

Comment (erlend-aasland) [reply Objects/typeobject.c]: Yes

Comment (erlend-aasland) [reply Objects/typeobject.c]: Hm, it should be %R, not %r

File: Objects/typeobject.c [hunk 3]
```diff
@@ -3947,8 +3951,8 @@ type_setattro(PyTypeObject *type, PyObject *name, PyObject *value)
     if (type->tp_flags & Py_TPFLAGS_IMMUTABLETYPE) {
         PyErr_Format(
             PyExc_TypeError,
-            "can't set attributes of built-in/extension type '%s'",
-            type->tp_name);
+            "cannot set '%R' attribute of immutable type '%s'",
```
Comment (vstinner) [inline Objects/typeobject.c]: '%R' would add too many quotes, no?

```suggestion
            "cannot set %R attribute of immutable type '%s'",
```

**Comment by vstinner** [comment]
I look the liberty of fixing ``'%R'`` before I noticed the test_descr crash :-(

**Comment by erlend-aasland** [comment]
> I look the liberty of fixing ``'%R'`` before I noticed the test_descr crash :-(

Thanks, Victor! I'm offline for an hour now. Feel free to take care of the fixes and land the PR :) Or I'll look at it later this afternoon.

File: Objects/typeobject.c (hunk 3 above)
Comment (erlend-aasland) [reply Objects/typeobject.c]: I was not aware of that :) Thanks!

**Comment by vstinner** [comment]
> Thanks, Victor! I'm offline for an hour now. Feel free to take care of the fixes and land the PR :) Or I'll look at it later this afternoon.

No problem, I pushed a fix for test_descr in your PR.

**Comment by erlend-aasland** [comment]
Thanks!

**Comment by vstinner** [comment]
Merged, thanks.

**Pull request merged**